### ISSUE TITLE
Cannot send new events into rooms with floating-point (and stringy?) power levels

### ISSUE THREAD

**Issue by DMRobertson** [matrix-org/synapse#14060]
Internal users: https://sentry.tools.element.io/organizations/element/issues/34032/?project=2&query=is%3Aunresolved&sort=freq&statsPeriod=14d

Sentry's report is terse here: it does not include local vars from the call stack.

```
Failed handle request via "RoomSendEventRestServlet": "<XForwardedForRequest at 0x7f8348f30f40 method='PUT' uri='/_matrix/client/r0/rooms/<room ID>/send/m.room.encrypted/<txn ID>' clientproto='HTTP/1.1' site='12101'>"
```

```
TypeError: argument 'notification_power_levels': 'float' object cannot be interpreted as an integer
  File "/home/synapse/src/synapse/http/server.py", line 306, in _async_render_wrapper
  File "/home/synapse/src/synapse/http/server.py", line 512, in _async_render
  File "/home/synapse/src/synapse/rest/client/room.py", line 353, in on_POST
  File "/home/synapse/src/synapse/handlers/message.py", line 1020, in create_and_send_nonmember_event
  File "/home/synapse/src/synapse/util/metrics.py", line 113, in measured_func
  File "/home/synapse/src/synapse/handlers/message.py", line 1382, in handle_new_client_event
  File "/home/synapse/env-pyston-poetry/lib/python3.8-pyston2.3/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
  File "/home/synapse/src/synapse/handlers/message.py", line 1436, in _persist_events
  File "/home/synapse/src/synapse/util/metrics.py", line 113, in measured_func
  File "/home/synapse/src/synapse/push/bulk_push_rule_evaluator.py", line 292, in action_for_event_by_user
```

The deepest call is

https://github.com/matrix-org/synapse/blob/b4ec4f5e71a87d5bdc840a4220dfd9a34c54c847/synapse/push/bulk_push_rule_evaluator.py#L292-L299

which invokes the new Rust mechanism for push stuff:

https://github.com/matrix-org/synapse/blob/ebd9e2dac6495a1857617d1a76c9259a988f8bb4/rust/src/push/evaluator.rs#L64-L90

I am guessing that

- the error message `'float' object cannot be interpreted as an integer` comes from Py03's translation layer. (#13838)
- the power levels event used in that room has floating power levels (and therefore is bogus, but not prohibited until room v10)
- people are unable to send messages in that room from Synapse  if they believe that power levels event to be in the current state

Will investigate a little more.

**Comment by DMRobertson** [comment]
```
matrix=> select json::json->'content'->'notifications' from current_state_events AS cse join event_json using(event_id) where cse.room_id = '<ROOM ID>' and cse.type = 'm.room.power_levels';
    ?column?    
════════════════
 {"room":100.0}
(1 row)

```

Right. Well that's not good.

**Comment by DMRobertson** [comment]
The event in question was created at `2020-12-29 08:52:34+00` according to its `origin_server_ts` so MSC3667 doesn't apply.

We now check that the power level values are integers in sufficiently new rooms.

https://github.com/matrix-org/synapse/blob/c2fe48a6ffb99f553f3eaecb8f15bcbedb58add0/synapse/event_auth.py#L783-L807

**Comment by DMRobertson** [comment]
No idea how we even got floats in the power levels content in the first place, given that https://github.com/matrix-org/synapse/blob/06df5d4250f54d5a95b0c90bfc9352ec6f02c520/synapse/events/utils.py#L662-L664 exists. But maybe it didn't at the time.

**Comment by DMRobertson** [comment]
Merged into develop in 44741aa85ba740df185549dc759a9e38ad51c16b.
